File: benchmarks/fft/java/Fft.java
public class Fft {
    static long nextPow2(long n) {
        long p = 1;
        while (p < n) p <<= 1;
        return p;
    }

    static void fft(double[] re, double[] im, boolean invert) {
        int n = re.length;
        int j = 0;
        for (int i = 1; i < n; i++) {
            int bit = n >> 1;
            while ((j & bit) != 0) {
                j ^= bit;
                bit >>= 1;
            }
            j ^= bit;
            if (i < j) {
                double tr = re[i]; re[i] = re[j]; re[j] = tr;
                double ti = im[i]; im[i] = im[j]; im[j] = ti;
            }
        }

        for (int length = 2; length <= n; length <<= 1) {
            double ang = 2.0 * Math.PI / length * (invert ? 1.0 : -1.0);
            double wlenRe = Math.cos(ang);
            double wlenIm = Math.sin(ang);
            for (int i = 0; i < n; i += length) {
                double wRe = 1.0, wIm = 0.0;
                for (int k = i; k < i + length / 2; k++) {
                    double uRe = re[k], uIm = im[k];
                    double vRe = re[k + length / 2] * wRe - im[k + length / 2] * wIm;
                    double vIm = re[k + length / 2] * wIm + im[k + length / 2] * wRe;
                    re[k] = uRe + vRe;
                    im[k] = uIm + vIm;
                    re[k + length / 2] = uRe - vRe;
                    im[k + length / 2] = uIm - vIm;
                    double nwRe = wRe * wlenRe - wIm * wlenIm;
                    double nwIm = wRe * wlenIm + wIm * wlenRe;
                    wRe = nwRe;
                    wIm = nwIm;
                }
            }
        }

        if (invert) {
            for (int i = 0; i < n; i++) {
                re[i] /= n;
                im[i] /= n;
            }
        }
    }

    public static void main(String[] args) {
        long requested = args.length > 0 ? Long.parseLong(args[0]) : 1_048_576L;
        int n = (int) nextPow2(requested);

        double[] re = new double[n];
        double[] im = new double[n];
        double[] orig = new double[n];
        for (int i = 0; i < n; i++) {
            double v = (i % 7) - 3.0;
            re[i] = v;
            orig[i] = v;
        }

        fft(re, im, false);
        fft(re, im, true);

        double maxErr = 0.0;
        for (int i = 0; i < n; i++) {
            double err = Math.max(Math.abs(re[i] - orig[i]), Math.abs(im[i]));
            if (err > maxErr) maxErr = err;
        }

        if (maxErr >= 1e-6) {
            System.err.println("self-check failed: roundtrip reconstruction error too large: " + maxErr);
            System.exit(1);
        }

        System.out.println(maxErr);
    }
}

File: benchmarks/fft/rust/fft.rs
use std::env;
use std::f64::consts::PI;

fn next_pow2(n: i64) -> i64 {
    let mut p: i64 = 1;
    while p < n {
        p <<= 1;
    }
    p
}

fn fft(re: &mut [f64], im: &mut [f64], invert: bool) {
    let n = re.len();
    let mut j = 0usize;
    for i in 1..n {
        let mut bit = n >> 1;
        while j & bit != 0 {
            j ^= bit;
            bit >>= 1;
        }
        j ^= bit;
        if i < j {
            re.swap(i, j);
            im.swap(i, j);
        }
    }

    let mut length = 2usize;
    while length <= n {
        let ang = 2.0 * PI / (length as f64) * (if invert { 1.0 } else { -1.0 });
        let wlen_re = ang.cos();
        let wlen_im = ang.sin();
        let mut i = 0;
        while i < n {
            let mut w_re = 1.0;
            let mut w_im = 0.0;
            for k in i..(i + length / 2) {
                let u_re = re[k];
                let u_im = im[k];
                let v_re = re[k + length / 2] * w_re - im[k + length / 2] * w_im;
                let v_im = re[k + length / 2] * w_im + im[k + length / 2] * w_re;
                re[k] = u_re + v_re;
                im[k] = u_im + v_im;
                re[k + length / 2] = u_re - v_re;
                im[k + length / 2] = u_im - v_im;
                let nw_re = w_re * wlen_re - w_im * wlen_im;
                let nw_im = w_re * wlen_im + w_im * wlen_re;
                w_re = nw_re;
                w_im = nw_im;
            }
            i += length;
        }
        length <<= 1;
    }

    if invert {
        for i in 0..n {
            re[i] /= n as f64;
            im[i] /= n as f64;
        }
    }
}

fn main() {
    let requested: i64 = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(1_048_576);
    let n = next_pow2(requested) as usize;

    let mut re = vec![0.0f64; n];
    let mut im = vec![0.0f64; n];
    let mut orig = vec![0.0f64; n];
    for i in 0..n {
        let v = (i % 7) as f64 - 3.0;
        re[i] = v;
        orig[i] = v;
    }

    fft(&mut re, &mut im, false);
    fft(&mut re, &mut im, true);

    let mut max_err = 0.0f64;
    for i in 0..n {
        let err = (re[i] - orig[i]).abs().max(im[i].abs());
        if err > max_err {
            max_err = err;
        }
    }

    assert!(max_err < 1e-6, "self-check failed: roundtrip reconstruction error too large: {}", max_err);

    println!("{:e}", max_err);
}
